# Credit Risk Project
## Phase 3 – Data Understanding

In [1]:
### 1. Import Libraries
import pandas as pd
from sqlalchemy import create_engine
from sqlalchemy.engine import URL

In [2]:
connection_url = URL.create(
    drivername="mysql+pymysql",
    username="root",
    password="Roxy@1994",   
    host="localhost",
    port=3306,
    database="credit_risk"
)

engine = create_engine(connection_url)

print("Connection Successful!")

Connection Successful!


In [3]:
### 3. load tables
application_train_df=pd.read_sql('select * from application_train',
                                 con=engine)




In [6]:
application_train_df.head()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
3,100006,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,...,0,0,0,0,,,,,,
4,100007,0,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0


In [7]:
### 3. verify data
application_train_df.shape

(307511, 122)

In [8]:
application_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(13), int64(41), object(68)
memory usage: 286.2+ MB


In [7]:
pd.set_option('display.float_format', '{:,.2f}'.format)

In [8]:
application_train_df.describe()

,SK_ID_CURR,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,...,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21
count,"307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00",...,"307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00","307,511.00"
mean,"278,180.52",0.08,0.42,"168,797.92","599,026.00","27,107.52","537,909.48",0.02,"-16,037.00","63,815.05",...,0.00,0.00,0.00,0.00,0.01,0.00,0.01,0.00,0.00,0.00
std,"102,790.18",0.27,0.72,"237,123.15","402,490.78","14,494.44","369,633.75",0.01,"4,363.99","141,275.77",...,0.00,0.06,0.05,0.03,0.10,0.02,0.09,0.02,0.02,0.02
min,"100,002.00",0.00,0.00,"25,650.00","45,000.00",0.00,0.00,0.00,"-25,229.00","-17,912.00",...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
25%,"189,145.50",0.00,0.00,"112,500.00","270,000.00","16,524.00","238,500.00",0.01,"-19,682.00","-2,760.00",...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
50%,"278,202.00",0.00,0.00,"147,150.00","513,531.00","24,903.00","450,000.00",0.02,"-15,750.00","-1,213.00",...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
75%,"367,142.50",0.00,1.00,"202,500.00","808,650.00","34,596.00","679,500.00",0.03,"-12,413.00",-289.00,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
max,"456,255.00",1.00,19.00,"117,000,000.00","4,050,000.00","258,025.50","4,050,000.00",0.07,"-7,489.00","365,243.00",...,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00


In [9]:
application_train_df.isnull().sum().sort_values(ascending=False).head(20)

SK_ID_CURR                      0
OBS_30_CNT_SOCIAL_CIRCLE        0
WALLSMATERIAL_MODE              0
TOTALAREA_MODE                  0
HOUSETYPE_MODE                  0
FONDKAPREMONT_MODE              0
NONLIVINGAREA_MEDI              0
NONLIVINGAPARTMENTS_MEDI        0
LIVINGAREA_MEDI                 0
LIVINGAPARTMENTS_MEDI           0
LANDAREA_MEDI                   0
FLOORSMIN_MEDI                  0
FLOORSMAX_MEDI                  0
ENTRANCES_MEDI                  0
ELEVATORS_MEDI                  0
COMMONAREA_MEDI                 0
YEARS_BUILD_MEDI                0
YEARS_BEGINEXPLUATATION_MEDI    0
BASEMENTAREA_MEDI               0
APARTMENTS_MEDI                 0
dtype: int64

In [10]:
application_train_df[['OWN_CAR_AGE', 'EXT_SOURCE_1']].head(10)

,OWN_CAR_AGE,EXT_SOURCE_1
0,,0.08303696739132256
1,,0.3112673113812225
2,26.0,
3,,
4,,
5,,
6,17.0,0.7747614130547695
7,8.0,
8,,0.5873340468730377
9,,


In [11]:
application_train_df['OWN_CAR_AGE'].unique()[:10]

array(['', '26.0', '17.0', '8.0', '23.0', '7.0', '14.0', '1.0', '3.0',
       '13.0'], dtype=object)

In [12]:
(application_train_df['OWN_CAR_AGE'] == '').sum()

np.int64(202929)

In [13]:
bureau_df=pd.read_sql('select * from bureau',con=engine)

In [15]:
bureau_df.head()

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY
0,215354,5714462,Closed,currency 1,-497,0,-153.0,-153.0,,0,"91,323.00",0.0,,0.00,Consumer credit,-131,
1,215354,5714463,Active,currency 1,-208,0,1075.0,,,0,"225,000.00",171342.0,,0.00,Credit card,-20,
2,215354,5714464,Active,currency 1,-203,0,528.0,,,0,"464,323.50",,,0.00,Consumer credit,-16,
3,215354,5714465,Active,currency 1,-203,0,,,,0,"90,000.00",,,0.00,Credit card,-16,
4,215354,5714466,Active,currency 1,-629,0,1197.0,,77674.5,0,"2,700,000.00",,,0.00,Consumer credit,-21,


In [16]:
bureau_df.shape

(1716428, 17)

In [17]:
bureau_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1716428 entries, 0 to 1716427
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_CURR              int64  
 1   SK_ID_BUREAU            int64  
 2   CREDIT_ACTIVE           object 
 3   CREDIT_CURRENCY         object 
 4   DAYS_CREDIT             int64  
 5   CREDIT_DAY_OVERDUE      int64  
 6   DAYS_CREDIT_ENDDATE     object 
 7   DAYS_ENDDATE_FACT       object 
 8   AMT_CREDIT_MAX_OVERDUE  object 
 9   CNT_CREDIT_PROLONG      int64  
 10  AMT_CREDIT_SUM          float64
 11  AMT_CREDIT_SUM_DEBT     object 
 12  AMT_CREDIT_SUM_LIMIT    object 
 13  AMT_CREDIT_SUM_OVERDUE  float64
 14  CREDIT_TYPE             object 
 15  DAYS_CREDIT_UPDATE      int64  
 16  AMT_ANNUITY             object 
dtypes: float64(2), int64(6), object(9)
memory usage: 222.6+ MB


In [19]:
bureau_balance_df=pd.read_sql('select * from bureau_balance',con=engine)

In [21]:
bureau_balance_df.head()

,SK_ID_BUREAU,MONTHS_BALANCE,STATUS
0,5715448,0,C
1,5715448,-1,C
2,5715448,-2,C
3,5715448,-3,C
4,5715448,-4,C


In [22]:
bureau_balance_df.shape

(27299925, 3)

In [29]:
bureau_balance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27299925 entries, 0 to 27299924
Data columns (total 3 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   SK_ID_BUREAU    int64 
 1   MONTHS_BALANCE  int64 
 2   STATUS          object
dtypes: int64(2), object(1)
memory usage: 624.8+ MB


In [24]:
previous_application_df=pd.read_sql('select * from previous_application',con=engine)

In [25]:
previous_application_df.head()

,SK_ID_PREV,SK_ID_CURR,NAME_CONTRACT_TYPE,AMT_ANNUITY,AMT_APPLICATION,AMT_CREDIT,AMT_DOWN_PAYMENT,AMT_GOODS_PRICE,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,...,NAME_SELLER_INDUSTRY,CNT_PAYMENT,NAME_YIELD_GROUP,PRODUCT_COMBINATION,DAYS_FIRST_DRAWING,DAYS_FIRST_DUE,DAYS_LAST_DUE_1ST_VERSION,DAYS_LAST_DUE,DAYS_TERMINATION,NFLAG_INSURED_ON_APPROVAL
0,2030495,271877,Consumer loans,"1,730.43","17,145.00","17,145.00",0.0,"17,145.00",SATURDAY,15,...,Connectivity,12.00,middle,POS mobile with interest,365243.0,-42.0,300.0,-42.0,-37.0,0.0
1,2802425,108129,Cash loans,"25,188.62","607,500.00","679,671.00",,"607,500.00",THURSDAY,11,...,XNA,36.00,low_action,Cash X-Sell: low,365243.0,-134.0,916.0,365243.0,365243.0,1.0
2,2523466,122040,Cash loans,"15,060.74","112,500.00","136,444.50",,"112,500.00",TUESDAY,11,...,XNA,12.00,high,Cash X-Sell: high,365243.0,-271.0,59.0,365243.0,365243.0,1.0
3,2819243,176158,Cash loans,"47,041.33","450,000.00","470,790.00",,"450,000.00",MONDAY,7,...,XNA,12.00,middle,Cash X-Sell: middle,365243.0,-482.0,-152.0,-182.0,-177.0,1.0
4,1784265,202054,Cash loans,"31,924.40","337,500.00","404,055.00",,"337,500.00",THURSDAY,9,...,XNA,24.00,high,Cash Street: high,,,,,,


In [28]:
previous_application_df.shape

(1670214, 37)

In [26]:
previous_application_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1670214 entries, 0 to 1670213
Data columns (total 37 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_PREV                   1670214 non-null  int64  
 1   SK_ID_CURR                   1670214 non-null  int64  
 2   NAME_CONTRACT_TYPE           1670214 non-null  object 
 3   AMT_ANNUITY                  1670214 non-null  float64
 4   AMT_APPLICATION              1670214 non-null  float64
 5   AMT_CREDIT                   1670214 non-null  float64
 6   AMT_DOWN_PAYMENT             1670214 non-null  object 
 7   AMT_GOODS_PRICE              1670214 non-null  float64
 8   WEEKDAY_APPR_PROCESS_START   1670214 non-null  object 
 9   HOUR_APPR_PROCESS_START      1670214 non-null  int64  
 10  FLAG_LAST_APPL_PER_CONTRACT  1670214 non-null  object 
 11  NFLAG_LAST_APPL_IN_DAY       1670214 non-null  int64  
 12  RATE_DOWN_PAYMENT            1670214 non-n

In [30]:
pos_cash_balance=pd.read_sql('select * from pos_cash_balance',con=engine)

In [32]:
pos_cash_balance.head()

,SK_ID_PREV,SK_ID_CURR,MONTHS_BALANCE,CNT_INSTALMENT,CNT_INSTALMENT_FUTURE,NAME_CONTRACT_STATUS,SK_DPD,SK_DPD_DEF
0,1803195,182943,-31,48.00,45.00,Active,0,0
1,1715348,367990,-33,36.00,35.00,Active,0,0
2,1784872,397406,-32,12.00,9.00,Active,0,0
3,1903291,269225,-35,48.00,42.00,Active,0,0
4,2341044,334279,-35,36.00,35.00,Active,0,0


In [33]:
pos_cash_balance.shape

(10001358, 8)

In [34]:
pos_cash_balance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10001358 entries, 0 to 10001357
Data columns (total 8 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   SK_ID_PREV             int64  
 1   SK_ID_CURR             int64  
 2   MONTHS_BALANCE         int64  
 3   CNT_INSTALMENT         float64
 4   CNT_INSTALMENT_FUTURE  float64
 5   NAME_CONTRACT_STATUS   object 
 6   SK_DPD                 int64  
 7   SK_DPD_DEF             int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 610.4+ MB


In [35]:
credit_card_df=pd.read_sql('select * from credit_card_balance',con=engine)

In [36]:
credit_card_df.head()

,SK_ID_PREV,SK_ID_CURR,MONTHS_BALANCE,AMT_BALANCE,AMT_CREDIT_LIMIT_ACTUAL,AMT_DRAWINGS_ATM_CURRENT,AMT_DRAWINGS_CURRENT,AMT_DRAWINGS_OTHER_CURRENT,AMT_DRAWINGS_POS_CURRENT,AMT_INST_MIN_REGULARITY,...,AMT_RECIVABLE,AMT_TOTAL_RECEIVABLE,CNT_DRAWINGS_ATM_CURRENT,CNT_DRAWINGS_CURRENT,CNT_DRAWINGS_OTHER_CURRENT,CNT_DRAWINGS_POS_CURRENT,CNT_INSTALMENT_MATURE_CUM,NAME_CONTRACT_STATUS,SK_DPD,SK_DPD_DEF
0,2562384,378907,-6,56.97,135000,0.00,877.50,0.00,877.50,"1,700.33",...,0.00,0.00,0.00,1,0.00,1.00,35.00,Active,0,0
1,2582071,363914,-1,"63,975.56",45000,"2,250.00","2,250.00",0.00,0.00,"2,250.00",...,"64,875.56","64,875.56",1.00,1,0.00,0.00,69.00,Active,0,0
2,1740877,371185,-7,"31,815.22",450000,0.00,0.00,0.00,0.00,"2,250.00",...,"31,460.08","31,460.08",0.00,0,0.00,0.00,30.00,Active,0,0
3,1389973,337855,-4,"236,572.11",225000,"2,250.00","2,250.00",0.00,0.00,"11,795.76",...,"233,048.97","233,048.97",1.00,1,0.00,0.00,10.00,Active,0,0
4,1891521,126868,-1,"453,919.46",450000,0.00,"11,547.00",0.00,"11,547.00","22,924.89",...,"453,919.46","453,919.46",0.00,1,0.00,1.00,101.00,Active,0,0


In [37]:
credit_card_df.shape

(3840312, 23)

In [38]:
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3840312 entries, 0 to 3840311
Data columns (total 23 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   SK_ID_PREV                  int64  
 1   SK_ID_CURR                  int64  
 2   MONTHS_BALANCE              int64  
 3   AMT_BALANCE                 float64
 4   AMT_CREDIT_LIMIT_ACTUAL     int64  
 5   AMT_DRAWINGS_ATM_CURRENT    float64
 6   AMT_DRAWINGS_CURRENT        float64
 7   AMT_DRAWINGS_OTHER_CURRENT  float64
 8   AMT_DRAWINGS_POS_CURRENT    float64
 9   AMT_INST_MIN_REGULARITY     float64
 10  AMT_PAYMENT_CURRENT         float64
 11  AMT_PAYMENT_TOTAL_CURRENT   float64
 12  AMT_RECEIVABLE_PRINCIPAL    float64
 13  AMT_RECIVABLE               float64
 14  AMT_TOTAL_RECEIVABLE        float64
 15  CNT_DRAWINGS_ATM_CURRENT    float64
 16  CNT_DRAWINGS_CURRENT        int64  
 17  CNT_DRAWINGS_OTHER_CURRENT  float64
 18  CNT_DRAWINGS_POS_CURRENT    float64
 19  CNT_INSTALMENT_MATURE

In [39]:
installment_payment_df=pd.read_sql('select * from installments_payments',con=engine)

In [40]:
installment_payment_df.head()

,SK_ID_PREV,SK_ID_CURR,NUM_INSTALMENT_VERSION,NUM_INSTALMENT_NUMBER,DAYS_INSTALMENT,DAYS_ENTRY_PAYMENT,AMT_INSTALMENT,AMT_PAYMENT
0,1054186,161674,1.00,6,"-1,180.00","-1,187.00","6,948.36","6,948.36"
1,1330831,151639,0.00,34,"-2,156.00","-2,156.00","1,716.53","1,716.53"
2,2085231,193053,2.00,1,-63.00,-63.00,"25,425.00","25,425.00"
3,2452527,199697,1.00,3,"-2,418.00","-2,426.00","24,350.13","24,350.13"
4,2714724,167756,1.00,2,"-1,383.00","-1,366.00","2,165.04","2,160.59"


In [41]:
installment_payment_df.shape

(13605401, 8)

In [43]:
installment_payment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13605401 entries, 0 to 13605400
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_PREV              int64  
 1   SK_ID_CURR              int64  
 2   NUM_INSTALMENT_VERSION  float64
 3   NUM_INSTALMENT_NUMBER   int64  
 4   DAYS_INSTALMENT         float64
 5   DAYS_ENTRY_PAYMENT      float64
 6   AMT_INSTALMENT          float64
 7   AMT_PAYMENT             float64
dtypes: float64(5), int64(3)
memory usage: 830.4 MB


### 3. Data quality Assesment 

In [9]:
application_train_df.dtypes

SK_ID_CURR                     int64
TARGET                         int64
NAME_CONTRACT_TYPE            object
CODE_GENDER                   object
FLAG_OWN_CAR                  object
                               ...  
AMT_REQ_CREDIT_BUREAU_DAY     object
AMT_REQ_CREDIT_BUREAU_WEEK    object
AMT_REQ_CREDIT_BUREAU_MON     object
AMT_REQ_CREDIT_BUREAU_QRT     object
AMT_REQ_CREDIT_BUREAU_YEAR    object
Length: 122, dtype: object

In [10]:
d_type_summary=pd.DataFrame({'column':application_train_df.columns,
                             'data_type':application_train_df.dtypes.values})
d_type_summary

,column,data_type
0,SK_ID_CURR,int64
1,TARGET,int64
2,NAME_CONTRACT_TYPE,object
3,CODE_GENDER,object
4,FLAG_OWN_CAR,object
...,...,...
117,AMT_REQ_CREDIT_BUREAU_DAY,object
118,AMT_REQ_CREDIT_BUREAU_WEEK,object
119,AMT_REQ_CREDIT_BUREAU_MON,object
120,AMT_REQ_CREDIT_BUREAU_QRT,object


In [12]:
pd.set_option('display.max_rows', None)

d_type_summary

,column,data_type
0,SK_ID_CURR,int64
1,TARGET,int64
2,NAME_CONTRACT_TYPE,object
3,CODE_GENDER,object
4,FLAG_OWN_CAR,object
5,FLAG_OWN_REALTY,object
6,CNT_CHILDREN,int64
7,AMT_INCOME_TOTAL,float64
8,AMT_CREDIT,float64
9,AMT_ANNUITY,float64


### Observation

- Dataset contains 307,511 rows and 122 columns.
- Most categorical variables are correctly stored as `object`.
- Most numerical variables are correctly stored as `int64` or `float64`.
- Several columns that appear to be numerical are currently stored as `object`.
- These columns will be investigated further during the data quality assessment before any type conversion is performed.

In [13]:
application_train_df["AMT_REQ_CREDIT_BUREAU_YEAR"].value_counts(dropna=False)

AMT_REQ_CREDIT_BUREAU_YEAR
0.0     71801
1.0     63405
2.0     50192
        41519
3.0     33628
4.0     20714
5.0     12052
6.0      6967
7.0      3869
8.0      2127
9.0      1096
11.0       31
12.0       30
10.0       22
13.0       19
14.0       10
17.0        7
15.0        6
19.0        4
18.0        4
16.0        3
25.0        1
23.0        1
22.0        1
21.0        1
20.0        1
Name: count, dtype: int64

The column `AMT_REQ_CREDIT_BUREAU_YEAR` contains 41,519 blank string values (`''`) representing missing data.

Because the column contains both numeric values and blank strings, Pandas inferred its data type as `object` instead of a numeric type.

This issue will be addressed during the Data Cleaning phase by replacing blank strings with `NaN` and converting the column to a numeric data type.

In [6]:
blank_summary=(application_train_df.select_dtypes(include='object').
    apply(lambda col:(col=='').sum()).sort_values(ascending=False))
blank_summary[blank_summary>0]
               

COMMONAREA_AVG                  214865
COMMONAREA_MODE                 214865
COMMONAREA_MEDI                 214865
NONLIVINGAPARTMENTS_MEDI        213514
NONLIVINGAPARTMENTS_MODE        213514
NONLIVINGAPARTMENTS_AVG         213514
FONDKAPREMONT_MODE              210295
LIVINGAPARTMENTS_MODE           210199
LIVINGAPARTMENTS_AVG            210199
LIVINGAPARTMENTS_MEDI           210199
FLOORSMIN_AVG                   208642
FLOORSMIN_MODE                  208642
FLOORSMIN_MEDI                  208642
YEARS_BUILD_AVG                 204488
YEARS_BUILD_MEDI                204488
YEARS_BUILD_MODE                204488
OWN_CAR_AGE                     202929
LANDAREA_AVG                    182590
LANDAREA_MEDI                   182590
LANDAREA_MODE                   182590
BASEMENTAREA_AVG                179943
BASEMENTAREA_MODE               179943
BASEMENTAREA_MEDI               179943
EXT_SOURCE_1                    173378
NONLIVINGAREA_MODE              169682
NONLIVINGAREA_MEDI       

In [7]:
numeric_missing = (
    application_train_df
    .select_dtypes(include=["int64", "float64"])
    .isna()
    .sum()
)

numeric_missing

SK_ID_CURR                     0
TARGET                         0
CNT_CHILDREN                   0
AMT_INCOME_TOTAL               0
AMT_CREDIT                     0
AMT_ANNUITY                    0
AMT_GOODS_PRICE                0
REGION_POPULATION_RELATIVE     0
DAYS_BIRTH                     0
DAYS_EMPLOYED                  0
DAYS_REGISTRATION              0
DAYS_ID_PUBLISH                0
FLAG_MOBIL                     0
FLAG_EMP_PHONE                 0
FLAG_WORK_PHONE                0
FLAG_CONT_MOBILE               0
FLAG_PHONE                     0
FLAG_EMAIL                     0
CNT_FAM_MEMBERS                0
REGION_RATING_CLIENT           0
REGION_RATING_CLIENT_W_CITY    0
HOUR_APPR_PROCESS_START        0
REG_REGION_NOT_LIVE_REGION     0
REG_REGION_NOT_WORK_REGION     0
LIVE_REGION_NOT_WORK_REGION    0
REG_CITY_NOT_LIVE_CITY         0
REG_CITY_NOT_WORK_CITY         0
LIVE_CITY_NOT_WORK_CITY        0
EXT_SOURCE_2                   0
OBS_30_CNT_SOCIAL_CIRCLE       0
DEF_30_CNT

In [10]:
application_train_csv=pd.read_csv(r'C:\Users\HomePC\Desktop\home_credit_dataset\application_train.csv')

In [11]:
application_train_csv.head()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
3,100006,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,...,0,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN
4,100007,0,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0


In [16]:
application_train_csv.shape

(307511, 122)

In [13]:
application_train_csv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


In [14]:
application_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(13), int64(41), object(68)
memory usage: 286.2+ MB


In [30]:
application_train_csv.isnull().sum()

SK_ID_CURR                        0
TARGET                            0
NAME_CONTRACT_TYPE                0
CODE_GENDER                       0
FLAG_OWN_CAR                      0
                              ...  
AMT_REQ_CREDIT_BUREAU_DAY     41519
AMT_REQ_CREDIT_BUREAU_WEEK    41519
AMT_REQ_CREDIT_BUREAU_MON     41519
AMT_REQ_CREDIT_BUREAU_QRT     41519
AMT_REQ_CREDIT_BUREAU_YEAR    41519
Length: 122, dtype: int64

In [20]:
numeric_missing = (
    application_train_csv
    .select_dtypes(include=["int64", "float64"])
    .isna()
    .sum()
)

numeric_missing[numeric_missing>0]

AMT_ANNUITY                       12
AMT_GOODS_PRICE                  278
OWN_CAR_AGE                   202929
CNT_FAM_MEMBERS                    2
EXT_SOURCE_1                  173378
                               ...  
AMT_REQ_CREDIT_BUREAU_DAY      41519
AMT_REQ_CREDIT_BUREAU_WEEK     41519
AMT_REQ_CREDIT_BUREAU_MON      41519
AMT_REQ_CREDIT_BUREAU_QRT      41519
AMT_REQ_CREDIT_BUREAU_YEAR     41519
Length: 61, dtype: int64

In [27]:
numeric_missing[numeric_missing>0].reset_index().sort_values(0,ascending=False)

,index,0
25,COMMONAREA_MODE,214865
39,COMMONAREA_MEDI,214865
11,COMMONAREA_AVG,214865
33,NONLIVINGAPARTMENTS_MODE,213514
19,NONLIVINGAPARTMENTS_AVG,213514
...,...,...
5,EXT_SOURCE_2,660
1,AMT_GOODS_PRICE,278
0,AMT_ANNUITY,12
3,CNT_FAM_MEMBERS,2


In [28]:
missing_report = pd.DataFrame({
    "Missing Count": application_train_csv.isnull().sum()
})

missing_report["Missing %"] = (
    missing_report["Missing Count"] / len(application_train_csv) * 100
).round(2)

missing_report = (
    missing_report[missing_report["Missing Count"] > 0]
    .sort_values("Missing Count", ascending=False)
)

missing_report

,Missing Count,Missing %
COMMONAREA_MEDI,214865,69.87
COMMONAREA_AVG,214865,69.87
COMMONAREA_MODE,214865,69.87
NONLIVINGAPARTMENTS_MEDI,213514,69.43
NONLIVINGAPARTMENTS_MODE,213514,69.43
...,...,...
EXT_SOURCE_2,660,0.21
AMT_GOODS_PRICE,278,0.09
AMT_ANNUITY,12,0.00
CNT_FAM_MEMBERS,2,0.00


In [31]:
missing_report.shape

(67, 2)

In [32]:
application_train_csv.isnull().sum().gt(0).sum()

np.int64(67)

In [40]:
missing_report['Missing_Category']=pd.cut(missing_report['Missing %'],
                                          bins=[0,5,20,50,100],
                                          labels=['0-5%','5-20%','20-50%',
                                                  '50-100%'],include_lowest=True)

In [41]:
missing_report["Missing_Category"].value_counts().sort_index()

Missing_Category
0-5%       10
5-20%       7
20-50%      9
50-100%    41
Name: count, dtype: int64

## Observation: Missing Value Assessment

- The `application_train` dataset contains **122 features**, of which **67 features contain missing values** and **55 features have no missing values**.
- Missing values are represented as **NaN** in the original dataset.
- Missing values are not uniformly distributed across the dataset; they are concentrated in specific groups of features.
- Housing-related features such as `COMMONAREA_*`, `NONLIVINGAPARTMENTS_*`, `LANDAREA_*`, `YEARS_BUILD_*`, and similar variables have the highest percentage of missing values, with several columns containing approximately **70% missing data**.
- External credit score variables (`EXT_SOURCE_1`, `EXT_SOURCE_2`, and `EXT_SOURCE_3`) also contain missing values.
- Most financial and demographic variables contain relatively few missing values compared to housing-related features.
- The distribution of missing values across the dataset is as follows:
  - **10 columns** contain **0–5%** missing values.
  - **7 columns** contain **5–20%** missing values.
  - **9 columns** contain **20–50%** missing values.
  - **41 columns** contain **50–100%** missing values.
- Missing value treatment will be performed during the **Data Cleaning** phase after evaluating the business relevance and predictive importance of each feature.

In [42]:
duplicate_count = application_train_csv.duplicated().sum()

print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 0


## Observation: Duplicate Record Assessment

- The `application_train` dataset contains **0 duplicate records**.
- Each record represents a unique loan application.
- No duplicate rows were identified in the dataset.
- Therefore, no duplicate records need to be removed during the Data Cleaning phase.

In [43]:
application_train_csv["SK_ID_CURR"].duplicated().sum()

np.int64(0)

### Primary Key Validation

- The `SK_ID_CURR` column contains **no duplicate values**.
- Each loan application is uniquely identified by `SK_ID_CURR`.
- This confirms that `SK_ID_CURR` can be treated as the unique identifier (Primary Key) for the `application_train` dataset.